### PR TITLE
Remove arrays from view

### PR DESCRIPTION
**Pull request by tbosch**
This improved the tree benchmark from 94ms down to 83ms in local tests.

### CODE REVIEW

File: modules/@angular/core/src/linker/element.ts [hunk 1]
```diff
@@ -22,8 +22,8 @@ import {ViewType} from './view_type';
  * that is needed for later instantiations.
  */
 export class AppElement {
-  public nestedViews: AppView<any>[] = null;
-  public componentView: AppView<any> = null;
+  public nestedViews: AppView<any>[];
+  public componentView: AppView<any>;
 
   public component: any;
   public componentConstructorViewQueries: QueryList<any>[];
```
Comment (tbosch) [inline modules/@angular/core/src/linker/element.ts]: emove this field

File: modules/@angular/compiler/src/view_compiler/compile_view.ts [hunk 2]
```diff
@@ -23,15 +23,29 @@ import {CompilePipe} from './compile_pipe';
 import {CompileQuery, addQueryToTokenMap, createQueryList} from './compile_query';
 import {getPropertyInView, getViewFactoryName} from './util';
 
+export enum CompileViewRootNodeType {
+  Node,
+  ViewContainer,
+  Projected
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/compile_view.ts]: Done

File: modules/@angular/core/src/linker/element.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/core/src/linker/element.ts]: Done

**Comment by tbosch** [comment]
Reviewed in person with @vicb

**Comment by vsavkin** [comment]
Merged https://github.com/angular/angular/commit/234c5599f10c33e743594b556b63bc6fdd87e7eb

**Pull request closed without merging**